Intermittant "We encountered an internal error. Please try again"

### Environment info
https://github.ibm.com/IBMSpectrumScale/hpo-core/issues/555
This is from HPO-core defect 
While running a cosbench test I get occasional errors. Sometimes it passes and sometimes it does not.

Errors occurred somewhere between these times
Start of run
Wed Mar 2 15:30:03 EST 2022
End of run
Wed Mar 2 15:34:19 EST 2022

A section from the cosbench log which shows the failure

[root@c83f1-dan4 ~]# 
From the Cosbench log  ================================================== stage: s1-prepare_GB ==================================================
---------------------------------- mission: MA4C534C68B, driver: app6 ----------------------------------
2022-03-02 15:30:02,008 [INFO] [Log4jLogManager] - will append log to file /root/cosbench/log/mission/MA4C534C68B.log
2022-03-02 15:30:02,235 [INFO] [NoneStorage] - performing PUT at /s5001b1/s5001o1_GB
2022-03-02 15:30:02,236 [INFO] [NoneStorage] - performing PUT at /s5001b1/s5001o3_GB
2022-03-02 15:30:02,236 [INFO] [NoneStorage] - performing PUT at /s5001b1/s5001o5_GB
2022-03-02 15:30:02,236 [INFO] [NoneStorage] - performing PUT at /s5001b1/s5001o2_GB
2022-03-02 15:30:02,236 [INFO] [NoneStorage] - performing PUT at /s5001b1/s5001o4_GB
2022-03-02 15:30:05,808 [INFO] [NoneStorage] - performing PUT at /s5001b2/s5001o4_GB
2022-03-02 15:30:06,344 [INFO] [NoneStorage] - performing PUT at /s5001b2/s5001o1_GB
2022-03-02 15:30:07,363 [WARN] [S3Storage] - below exception encountered when creating object s5001o2_GB at s5001b1: Failed to reset the request input stream;  If the request involves an input stream, the maximum stream buffer size can be configured via request.getRequestClientOptions().setReadLimit(int)
2022-03-02 15:30:07,365 [INFO] [NoneStorage] - performing PUT at /s5001b2/s5001o2_GB
2022-03-02 15:30:10,220 [INFO] [NoneStorage] - performing PUT at /s5001b3/s5001o1_GB
2022-03-02 15:30:10,269 [INFO] [NoneStorage] - performing PUT at /s5001b2/s5001o3_GB
2022-03-02 15:30:11,647 [INFO] [NoneStorage] - performing PUT at /s5001b2/s5001o5_GB
2022-03-02 15:30:14,219 [INFO] [NoneStorage] - performing PUT at /s5001b3/s5001o3_GB
2022-03-02 15:30:16,229 [INFO] [NoneStorage] - performing PUT at /s5001b3/s5001o4_GB
2022-03-02 15:30:18,461 [INFO] [NoneStorage] - performing PUT at /s5001b3/s5001o2_GB
2022-03-02 15:30:19,597 [INFO] [NoneStorage] - performing PUT at /s5001b4/s5001o4_GB
2022-03-02 15:30:19,948 [WARN] [S3Storage] - below exception encountered when creating object s5001o3_GB at s5001b3: Failed to reset the request input stream;  If the request involves an input stream, the maximum stream buffer size can be configured via request.getRequestClientOptions().setReadLimit(int)
2022-03-02 15:30:19,948 [INFO] [NoneStorage] - performing PUT at /s5001b4/s5001o3_GB
2022-03-02 15:30:21,821 [INFO] [NoneStorage] - performing PUT at /s5001b4/s5001o1_GB
2022-03-02 15:30:23,736 [INFO] [NoneStorage] - performing PUT at /s5001b3/s5001o5_GB
2022-03-02 15:30:31,439 [INFO] [NoneStorage] - performing PUT at /s5001b4/s5001o2_GB
2022-03-02 15:30:39,427 [INFO] [NoneStorage] - performing PUT at /s5001b4/s5001o5_GB

Picking out this error to follow: 

2022-03-02 15:30:07,363 [WARN] [S3Storage] - below exception encountered when creating object s5001o2_GB at s5001b1: Failed to reset the request input stream;  If the request involves an input stream, the maximum stream buffer size can be configured via 

Looking for that error in the noobaa log:

[root@c83f1-infa internal]# oc get pod -o wide | grep -i  endpoint | grep dan3 | awk '{print $1}
> '| xargs -ti oc logs {} | grep 20:30:07 | grep s5001o2
oc logs noobaa-endpoint-749785777b-2fzxd
oc logs noobaa-endpoint-749785777b-2g7s4
oc logs noobaa-endpoint-749785777b-6jp42
oc logs noobaa-endpoint-749785777b-6wgcc
oc logs noobaa-endpoint-749785777b-7sx8t
oc logs noobaa-endpoint-749785777b-btkhh
oc logs noobaa-endpoint-749785777b-gsrx9
oc logs noobaa-endpoint-749785777b-spp8b
oc logs noobaa-endpoint-749785777b-wf4p4
oc logs noobaa-endpoint-749785777b-xsqxc
Mar-2 20:30:07.379 [Endpoint/14] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/s5001b1/s5001o2_GB</Resource><RequestId>l0a0j66s-amgalv-n9a</RequestId></Error> PUT /s5001b1/s5001o2_GB {"host":"metallb","authorization":"AWS lDT36NrV8XNBMg0p8ESh:DNkEsP3hIfzKYQSH5sqT2iKjKW8=","user-agent":"aws-sdk-java/1.10.76 Linux/4.18.0-240.el8.x86_64 OpenJDK_64-Bit_Server_VM/25.265-b01/1.8.0_265","amz-sdk-invocation-id":"f7e25c94-f1b5-436a-bc64-97c730ead3fa","amz-sdk-retry":"0/0/","date":"Wed, 02 Mar 2022 20:30:02 GMT","content-type":"application/octet-stream","content-length":"3000000000","connection":"Keep-Alive"} TypeError: callback is not a function
Mar-2 20:30:07.427 [Endpoint/14]    [L0] core.endpoint.s3.ops.s3_put_object:: PUT OBJECT s5001b2 s5001o2_GB

 ```
root@c83f1-infa internal]# noobaa status
INFO[0000] CLI version: 5.9.0
INFO[0000] noobaa-image: noobaa/noobaa-core:5.9_nsfsfixes-20220215
INFO[0000] operator-image: quay.io/rhceph-dev/odf4-mcg-rhel8-operator@sha256:773dbaded46fa0a024c28a92f98f4ad64d370011ed1405f2b16f39b9258eb6b2
INFO[0000] noobaa-db-image: quay.io/rhceph-dev/rhel8-postgresql-12@sha256:da0b8d525b173ef472ff4c71fae60b396f518860d6313c4f3287b844aab6d622
INFO[0000] Namespace: openshift-storage
INFO[0000]
INFO[0000] CRD Status:
INFO[0000] ✅ Exists: CustomResourceDefinition "noobaas.noobaa.io"
INFO[0000] ✅ Exists: CustomResourceDefinition "backingstores.noobaa.io"
INFO[0000] ✅ Exists: CustomResourceDefinition "namespacestores.noobaa.io"
INFO[0000] ✅ Exists: CustomResourceDefinition "bucketclasses.noobaa.io"
INFO[0000] ✅ Exists: CustomResourceDefinition "noobaaaccounts.noobaa.io"
INFO[0000] ✅ Exists: CustomResourceDefinition "objectbucketclaims.objectbucket.io"
INFO[0000] ✅ Exists: CustomResourceDefinition "objectbuckets.objectbucket.io"
INFO[0000]
INFO[0000] Operator Status:
INFO[0000] ✅ Exists: Namespace "openshift-storage"
INFO[0000] ✅ Exists: ServiceAccount "noobaa"
INFO[0000] ✅ Exists: ServiceAccount "noobaa-endpoint"
INFO[0000] ✅ Exists: Role "mcg-operator.v4.9.3-noobaa-7fdbb75fd7"
INFO[0000] ✅ Exists: Role "mcg-operator.v4.9.3-noobaa-endpoint-65854bfccb"
INFO[0000] ✅ Exists: RoleBinding "mcg-operator.v4.9.3-noobaa-7fdbb75fd7"
INFO[0000] ✅ Exists: RoleBinding "mcg-operator.v4.9.3-noobaa-endpoint-65854bfccb"
INFO[0000] ✅ Exists: ClusterRole "mcg-operator.v4.9.3-644558fccd"
INFO[0000] ✅ Exists: ClusterRoleBinding "mcg-operator.v4.9.3-644558fccd"
INFO[0000] ⬛ (Optional) Not Found: ValidatingWebhookConfiguration "admission-validation-webhook"
INFO[0000] ⬛ (Optional) Not Found: Secret "admission-webhook-secret"
INFO[0000] ⬛ (Optional) Not Found: Service "admission-webhook-service"
INFO[0000] ✅ Exists: Deployment "noobaa-operator"
INFO[0000]
INFO[0000] System Wait Ready:
INFO[0000] ✅ System Phase is "Ready".
INFO[0000]
INFO[0000]
INFO[0000] System Status:
INFO[0000] ✅ Exists: NooBaa "noobaa"
INFO[0000] ✅ Exists: StatefulSet "noobaa-core"
INFO[0000] ✅ Exists: ConfigMap "noobaa-config"
INFO[0000] ✅ Exists: Service "noobaa-mgmt"
INFO[0000] ✅ Exists: Service "s3"
INFO[0000] ✅ Exists: Secret "noobaa-db"
INFO[0000] ✅ Exists: ConfigMap "noobaa-postgres-config"
INFO[0000] ✅ Exists: ConfigMap "noobaa-postgres-initdb-sh"
INFO[0000] ✅ Exists: StatefulSet "noobaa-db-pg"
INFO[0000] ✅ Exists: Service "noobaa-db-pg"
INFO[0000] ✅ Exists: Secret "noobaa-server"
INFO[0000] ✅ Exists: Secret "noobaa-operator"
INFO[0000] ✅ Exists: Secret "noobaa-endpoints"
INFO[0000] ✅ Exists: Secret "noobaa-admin"
INFO[0000] ✅ Exists: StorageClass "openshift-storage.noobaa.io"
INFO[0000] ✅ Exists: BucketClass "noobaa-default-bucket-class"
INFO[0000] ✅ Exists: Deployment "noobaa-endpoint"
INFO[0000] ✅ Exists: HorizontalPodAutoscaler "noobaa-endpoint"
INFO[0000] ✅ (Optional) Exists: BackingStore "noobaa-default-backing-store"
INFO[0001] ⬛ (Optional) Not Found: CredentialsRequest "noobaa-aws-cloud-creds"
INFO[0001] ⬛ (Optional) Not Found: CredentialsRequest "noobaa-azure-cloud-creds"
INFO[0001] ⬛ (Optional) Not Found: Secret "noobaa-azure-container-creds"
INFO[0001] ⬛ (Optional) Not Found: Secret "noobaa-gcp-bucket-creds"
INFO[0001] ⬛ (Optional) Not Found: CredentialsRequest "noobaa-gcp-cloud-creds"
INFO[0001] ✅ (Optional) Exists: PrometheusRule "noobaa-prometheus-rules"
INFO[0001] ✅ (Optional) Exists: ServiceMonitor "noobaa-mgmt-service-monitor"
INFO[0001] ✅ (Optional) Exists: ServiceMonitor "s3-service-monitor"
INFO[0001] ✅ (Optional) Exists: Route "noobaa-mgmt"
INFO[0001] ✅ (Optional) Exists: Route "s3"
INFO[0001] ✅ Exists: PersistentVolumeClaim "db-noobaa-db-pg-0"
INFO[0001] ✅ System Phase is "Ready"
INFO[0001] ✅ Exists: "noobaa-admin"

#------------------#
#- Mgmt Addresses -#
#------------------#

ExternalDNS : [https://noobaa-mgmt-openshift-storage.apps.ocp4.pokprv.stglabs.ibm.com]
ExternalIP : []
NodePorts : [https://10.28.20.46:32659]
InternalDNS : [https://noobaa-mgmt.openshift-storage.svc:443]
InternalIP : [https://172.30.179.247:443]
PodPorts : [https://10.128.3.154:8443]

#--------------------#
#- Mgmt Credentials -#
#--------------------#

email : [admin@noobaa.io](mailto:admin@noobaa.io)
password : BWb0l3hATSA1+BUyJrShrw==

#----------------#
#- S3 Addresses -#
#----------------#

ExternalDNS : [https://s3-openshift-storage.apps.ocp4.pokprv.stglabs.ibm.com]
ExternalIP : []
NodePorts : [https://10.28.20.47:30905 https://10.28.20.46:30905 https://10.28.20.47:30905 https://10.28.20.45:30905 https://10.28.20.45:30905 https://10.28.20.45:30905 https://10.28.20.47:30905 https://10.28.20.45:30905 https://10.28.20.46:30905 https://10.28.20.46:30905 https://10.28.20.47:30905 https://10.28.20.45:30905 https://10.28.20.46:30905 https://10.28.20.45:30905 https://10.28.20.46:30905 https://10.28.20.47:30905 https://10.28.20.46:30905 https://10.28.20.45:30905 https://10.28.20.47:30905 https://10.28.20.45:30905 https://10.28.20.46:30905 https://10.28.20.47:30905 https://10.28.20.47:30905 https://10.28.20.46:30905 https://10.28.20.45:30905 https://10.28.20.45:30905 https://10.28.20.47:30905 https://10.28.20.47:30905 https://10.28.20.46:30905 https://10.28.20.46:30905]
InternalDNS : [https://s3.openshift-storage.svc:443]
InternalIP : [https://172.30.167.87:443]
PodPorts : [https://10.128.1.89:6443 https://10.128.2.173:6443 https://10.128.1.90:6443 https://10.128.4.152:6443 https://10.128.4.154:6443 https://10.128.4.150:6443 https://10.128.1.92:6443 https://10.128.4.147:6443 https://10.128.2.178:6443 https://10.128.2.176:6443 https://10.128.1.93:6443 https://10.128.4.153:6443 https://10.128.2.182:6443 https://10.128.4.148:6443 https://10.128.2.175:6443 https://10.128.1.87:6443 https://10.128.2.177:6443 https://10.128.4.149:6443 https://10.128.1.88:6443 https://10.128.4.155:6443 https://10.128.2.174:6443 https://10.128.1.95:6443 https://10.128.1.86:6443 https://10.128.2.181:6443 https://10.128.4.151:6443 https://10.128.4.146:6443 https://10.128.1.94:6443 https://10.128.1.91:6443 https://10.128.2.179:6443 https://10.128.2.180:6443]

#------------------#
#- S3 Credentials -#
#------------------#

AWS_ACCESS_KEY_ID : xbtyDghqOmoE3RDoWULe
AWS_SECRET_ACCESS_KEY : q9BE7S7KTVZX8N/A5LB9sh9Kv1dSOB/d1P2QjEwg

#------------------#
#- Backing Stores -#
#------------------#

NAME TYPE TARGET-BUCKET PHASE AGE
noobaa-default-backing-store pv-pool Ready 6d1h28m2s

#--------------------#
#- Namespace Stores -#
#--------------------#

NAME TYPE TARGET-BUCKET PHASE AGE
noobaa-s3res-3777249441 nsfs Ready 6d1h27m56s

#------------------#
#- Bucket Classes -#
#------------------#

NAME PLACEMENT NAMESPACE-POLICY QUOTA PHASE AGE
noobaa-default-bucket-class {"tiers":[{"backingStores":["noobaa-default-backing-store"]}]} null null Ready 6d1h28m2s

#-------------------#
#- NooBaa Accounts -#
#-------------------#

No noobaa accounts found.

#-----------------#
#- Bucket Claims -#
#-----------------#

No OBCs found.


Must gathers are in box note https://ibm.ent.box.com/folder/145794528783?s=uueh7fp424vxs2bt4ndrnvh7uusgu6tocd 
Could not get oc adm must-gather --image=registry.redhat.io/ocs4/ocs-must-gather-rhel8
b/c of this error:
must-gather-j669f] POD 2022-03-02T21:43:59.287340940Z collecting dump cephblockpools
[must-gather-j669f] POD 2022-03-02T21:43:59.508140313Z collecting dump cephclusters
[must-gather-j669f] POD 2022-03-02T21:43:59.738039077Z collecting dump cephfilesystems
[must-gather-j669f] POD 2022-03-02T21:43:59.978852540Z collecting dump cephobjectstores
[must-gather-j669f] POD 2022-03-02T21:44:00.207898724Z collecting dump cephobjectstoreusers
[must-gather-j669f] POD 2022-03-02T21:44:01.724486444Z Error from server (NotFound): pods "must-gather-j669f-helper" not found
[must-gather-j669f] POD 2022-03-02T21:44:01.738185148Z waiting for helper pod to come up in openshift-storage namespace. Retrying 1
[must-gather-j669f] POD 2022-03-02T21:44:06.926669468Z Error from server (NotFound): pods "must-gather-j669f-helper" not found
[must-gather-j669f] POD 2022-03-02T21:44:06.961050391Z waiting for helper pod to come up in openshift-storage namespace. Retrying 2
[must-gather-j669f] POD 2022-03-02T21:44:12.175332535Z Error from server (NotFound): pods "must-gather-j66

### Your environment
[root@c83f1-infa internal]# oc get csv
NAME                  DISPLAY                       VERSION   REPLACES              PHASE
mcg-operator.v4.9.3   NooBaa Operator               4.9.3     mcg-operator.v4.9.2   Succeeded
ocs-operator.v4.9.3   OpenShift Container Storage   4.9.3     ocs-operator.v4.9.2   Succeeded
odf-operator.v4.9.3   OpenShift Data Foundation     4.9.3     odf-operator.v4.9.2   Succeeded
[root@c83f1-infa internal]#

POK Bare Metal, Real Fast cluster
30 enpoint pods
Md5 disabled
Default CPU/memory on endpoint pods
1 account being tested even though there are 100 accounts.  There are 10 buckets for this account
128 workers in Cosbench
4 app nodes but looks like only 2 were used

### Steps to reproduce
Cosbench xml
<?xml version="1.0" encoding="UTF-8" ?>
<workload name="GBrangesizeobj" description="GB range read and write workload">

  <storage type="s3" config="accesskey=lDT36NrV8XNBMg0p8ESh;secretkey=TZDzPCqqbBhjs1nzQX0Y0fK27F2ANCTP81iQjjqj;endpoint=http://metallb:80;path_style_access=true" />

  <workflow>

   <workstage name="prepare_GB">
      <work name="prepare_GB" workers="128" interval="10" rampup="0" type="prepare" ratio="100" config="cprefix=s5001b;oprefix=s5001o;osuffix=_GB;containers=r(1,4);objects=r(1,5);sizes=r(1,3)GB"/>
    </workstage>


   <workstage name="rangewritereaddelete">
      <work name="GBrangewritereaddelete" workers="128" interval="10" rampup="0" runtime="1800">
        <operation type="write" ratio="20" config="cprefix=s5001b;oprefix=s5001o;osuffix=_GB;containers=r(1,4);objects=r(7,9);sizes=r(1,3)GB"/>
        <operation type="read" ratio="60" config="cprefix=s5001b;oprefix=s5001o;osuffix=_GB;containers=r(1,4);objects=r(1,3)"/>
        <operation type="delete" ratio="20" config="cprefix=s5001b;oprefix=s5001o;osuffix=_GB;containers=r(1,4);objects=r(4,6)"/>
      </work>
    </workstage>


   <workstage name="cleanup_GB">
      <work name="cleanup_GB" workers="128" interval="10" rampup="0" type="cleanup" ratio="100" config="cprefix=s5001b;oprefix=s5001o;osuffix=_GB;containers=r(1,4);objects=r(1,9)"/>
    </workstage>

  </workflow>

</workload>


### Expected behaviour

Should be able to write objects consistently

### Actual behavior
1.

### Expected behavior
1.

### Steps to reproduce
1.

### More information - Screenshots / Logs / Other output

[](https://media.github.ibm.com/user/169988/files/756e0a00-ac36-11ec-9414-a4cbc6da20c4)I was able to recreate with nsfs logging and a very short cosbench xml file that only does the prepare stage:

[root@c83f1-dan4 ~]# cat /root/cosbench/workloads/RW_workloads/rand_read_write_delete_GB_1hour/s5003_GB_random_size_num_of_objects.xml
<?xml version="1.0" encoding="UTF-8" ?>
<workload name="GBrandomsizenumobj" description="GBrandom read and write workload">
<storage type="s3" config="accesskey=4vv03cSdThLqVhHXzVAs;secretkey=SthtDwpJr6UM6d3rcEw3KDl1n+sNQJUp4Nx4CAvF;path_style_access=true;endpoint=http://metallb:80"/>

<workflow>

 <workstage name="prepare_rGB">
    <work name="prepare_rGB" workers="256" interval="10" rampup="0" type="prepare" ratio="100" config="cprefix=s5003b;oprefix=s5003o;osuffix=_rGB;containers=r(1,4);objects=r(1,2);sizes=r(2,3)GB"/>
  </workstage>


</workflow>
</workload>

The run is started

[root@c83f1-dan4 ~]# date ; sh /root/cosbench/cli.sh submit /root/cosbench/workloads/RW_workloads/rand_read_write_delete_GB_1hour/s5003_GB_random_size_num_of_objects.xml
Fri Mar 25 13:09:48 EDT 2022
Accepted with ID: w899

It completes itself and there is no CTL-C
Screen Shot 2022-03-25 at 12 12 32 PM

When I look at the files on the storage cluster:

From the storage cluster
[root@c83f1-ems s5003]# find ./s500* -type f
./s5003b1/s5003o2_rGB
./s5003b1/s5003o1_rGB
./s5003b2/s5003o2_rGB
./s5003b2/s5003o1_rGB
./s5003b3/s5003o2_rGB
./s5003b3/s5003o1_rGB
./s5003b4/.noobaa-nsfs_622115d4c8206400294c9c6e/uploads/c47886f6-b3c9-4b40-817c-a32055aa2049
./s5003b4/s5003o1_rGB
[root@c83f1-ems s5003]#

Oc logs shows this error:

[root@c83f1-infa 899]# date; oc logs deployment/noobaa-endpoint --all-containers=true --prefix=true --max-log-requests=10  --since=25m  | grep PUT | grep -i error
Fri Mar 25 13:20:05 EDT 2022
Found 3 pods, using pod/noobaa-endpoint-677c67c99c-kkjgh
[pod/noobaa-endpoint-677c67c99c-kkjgh/endpoint] Mar-25 17:10:56.145 [Endpoint/15] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><Resource>/s5003b4/s5003o2_rGB</Resource><RequestId>l16ojg57-fw67mb-o1c</RequestId></Error> PUT /s5003b4/s5003o2_rGB {"host":"metallb","authorization":"AWS 4vv03cSdThLqVhHXzVAs:CAyk9U9UlMFTBxYfLV4SGQbwRxE=","user-agent":"aws-sdk-java/1.10.76 Linux/4.18.0-240.el8.x86_64 OpenJDK_64-Bit_Server_VM/25.265-b01/1.8.0_265","amz-sdk-invocation-id":"da1f47d2-cdcd-428f-b118-8f2e52703692","amz-sdk-retry":"0/0/","date":"Fri, 25 Mar 2022 17:10:43 GMT","content-type":"application/octet-stream","content-length":"3000000000","connection":"Keep-Alive"} TypeError: callback is not a function

The OCP is using 3 endpoint pods and default cpu (2) memory (3G). Also Cosbench has 4 drivers but it looks like according to the cosbench log only 1 driver was used.

The Cosbench log ================================================== stage: s1-prepare_rGB ==================================================
---------------------------------- mission: MCC20E3E062, driver: app6 ----------------------------------
2022-03-25 13:09:48,027 [INFO] [Log4jLogManager] - will append log to file /root/cosbench/log/mission/MCC20E3E062.log
2022-03-25 13:09:48,297 [INFO] [NoneStorage] - performing PUT at /s5003b1/s5003o1_rGB
2022-03-25 13:09:48,297 [INFO] [NoneStorage] - performing PUT at /s5003b1/s5003o2_rGB
2022-03-25 13:10:05,216 [INFO] [NoneStorage] - performing PUT at /s5003b2/s5003o1_rGB
2022-03-25 13:10:10,715 [INFO] [NoneStorage] - performing PUT at /s5003b2/s5003o2_rGB
2022-03-25 13:10:16,604 [INFO] [NoneStorage] - performing PUT at /s5003b3/s5003o1_rGB
2022-03-25 13:10:27,854 [INFO] [NoneStorage] - performing PUT at /s5003b3/s5003o2_rGB
2022-03-25 13:10:28,703 [INFO] [NoneStorage] - performing PUT at /s5003b4/s5003o1_rGB
2022-03-25 13:10:43,417 [INFO] [NoneStorage] - performing PUT at /s5003b4/s5003o2_rGB
2022-03-25 13:10:55,982 [WARN] [S3Storage] - below exception encountered when creating object s5003o2_rGB at s5003b4: Failed to reset the request input stream;  If the request involves an input stream, the maximum stream buffer size can be configured via request.getRequestClientOptions().setReadLimit(int)
---------------------------------- mission: M2C20E3E06F, driver: app7 ----------------------------------
2022-03-25 13:09:48,026 [INFO] [Log4jLogManager] - will append log to file /root/cosbench/log/mission/M2C20E3E06F.log
---------------------------------- mission: M4C20E3E0E9, driver: app8 ----------------------------------
2022-03-25 13:09:48,018 [INFO] [Log4jLogManager] - will append log to file /root/cosbench/log/mission/M4C20E3E0E9.log
---------------------------------- mission: MCC20E3AFB3, driver: dan4 ----------------------------------
2022-03-25 13:09:47,023 [INFO] [Log4jLogManager] - will append log to file /root/cosbench/log/mission/MCC20E3AFB3.log

I grabbed a couple of noobaa diagnose and must-gather b/c I did not know when logs would wrap.
They are in box note:
https://ibm.ent.box.com/folder/145794528783?s=uueh7fp424vxs2bt4ndrnvh7uusgu6tocd  
must-gather-6896-run899.tarz

I hope that this recreate provides everything that is needed

--- a/config.js
+++ b/config.js
@@ -556,6 +556,7 @@ assert(config.NAMESPACE_CACHING.DEFAULT_BLOCK_SIZE > config.INLINE_MAX_SIZE);
 
 config.NSFS_BUF_SIZE = 8 * 1024 * 1024;
 config.NSFS_BUF_POOL_MEM_LIMIT = config.BUFFERS_MEM_LIMIT;
+config.NSFS_DEFAULT_IOV_MAX = 1024; // see IOV_MAX in https://man7.org/linux/man-pages/man0/limits.h.0p.html
 
 // the temporary path for uploads and other internal files
 config.NSFS_TEMP_DIR_NAME = '.noobaa-nsfs';

--- a/src/native/fs/fs_napi.cpp
+++ b/src/native/fs/fs_napi.cpp
@@ -20,6 +20,7 @@
 #include <sys/xattr.h>
 #include <vector>
 #include <stdlib.h>
+#include <limits.h>
 
 #ifdef __APPLE__
     #include <sys/param.h>
@@ -1412,6 +1413,7 @@ fs_napi(Napi::Env env, Napi::Object exports)
     exports_fs["S_IFLNK"] = Napi::Number::New(env, S_IFLNK);
     exports_fs["DT_DIR"] = Napi::Number::New(env, DT_DIR);
     exports_fs["DT_LNK"] = Napi::Number::New(env, DT_LNK);
+    exports_fs["PLATFORM_IOV_MAX"] = Napi::Number::New(env, IOV_MAX);
 
     exports_fs["set_debug_level"] = Napi::Function::New(env, set_debug_level);
 

--- a/src/test/unit_tests/index.js
+++ b/src/test/unit_tests/index.js
@@ -4,7 +4,6 @@
 
 
 const coretest = require('./coretest');
-const config = require('../../../config');
 coretest.setup({ incomplete_rpc_coverage: 'show' });
 
 // ---------------------------------------
@@ -60,7 +59,7 @@ require('./test_mirror_writer');
 require('./test_namespace_fs');
 require('./test_ns_list_objects');
 require('./test_bucketspace');
-if (config.NSFS_CALCULATE_MD5) require('./test_chunk_fs');
+require('./test_chunk_fs');
 
 // // SERVERS
 require('./test_agent');

--- a/src/test/unit_tests/test_chunk_fs.js
+++ b/src/test/unit_tests/test_chunk_fs.js
@@ -19,4 +19,16 @@ mocha.describe('ChunkFS', function() {
         self.timeout(RUN_TIMEOUT);
         await chunk_fs_hashing.file_target();
     });
+
+    mocha.it('Concurrent ChunkFS with file target - produce num_chunks > 1024 && total_chunks_size < config.NSFS_BUF_SIZE', async function() {
+        const self = this;
+        self.timeout(RUN_TIMEOUT);
+        // The goal of this test is to produce num_chunks > 1024 && total_chunks_size < config.NSFS_BUF_SIZE
+        // so we will flush buffers because of reaching max num of buffers and not because we reached the max NSFS buf size
+        // chunk size = 100, num_chunks = (10 * 1024 * 1024)/100 < 104587, 104587 = num_chunks > 1024 
+        // chunk size = 100, total_chunks_size after having 1024 chunks is = 100 * 1024 < config.NSFS_BUF_SIZE
+        const chunk_size = 100;
+        const parts_s = 50;
+        await chunk_fs_hashing.file_target(chunk_size, parts_s);
+    });
 });

--- a/src/tools/chunk_fs_hashing.js
+++ b/src/tools/chunk_fs_hashing.js
@@ -13,8 +13,8 @@ const fs = require('fs');
 const argv = require('minimist')(process.argv);
 const path = require('path');
 
-const XATTR = argv.fsync || true;
-const FSYNC = argv.xattr || true;
+const XATTR = argv.xattr || true;
+const FSYNC = argv.fsync || true;
 const PARTS = Number(argv.parts) || 1000;
 const CONCURRENCY = Number(argv.concurrency) || 20;
 const CHUNK = Number(argv.chunk) || 16 * 1024;
@@ -91,14 +91,16 @@ async function hash_target() {
             `DataOriginMD5=${content_md5}`,
         );
         assert.strictEqual(content_md5, write_hash);
-        assert.strictEqual(chunk_fs.digest, content_md5);
-        assert.strictEqual(chunk_fs.digest, write_hash);
+        if (config.NSFS_CALCULATE_MD5) {
+            assert.strictEqual(chunk_fs.digest, content_md5);
+            assert.strictEqual(chunk_fs.digest, write_hash);
+        }
     });
 }
 
-async function file_target() {
+async function file_target(chunk_size = CHUNK, parts = PARTS) {
     fs.mkdirSync(F_PREFIX);
-    await P.map_with_concurrency(CONCURRENCY, Array(PARTS).fill(), async () => {
+    await P.map_with_concurrency(CONCURRENCY, Array(parts).fill(), async () => {
         let target_file;
         const data = crypto.randomBytes(PART_SIZE);
         const content_md5 = crypto.createHash('md5').update(data).digest('hex');
@@ -107,8 +109,8 @@ async function file_target() {
             target_file = await nb_native().fs.open(DEFAULT_FS_CONFIG, F_TARGET, 'w', get_umasked_mode(config.BASE_MODE_FILE));
             // Using async generator function in order to push data in small chunks
             const source_stream = stream.Readable.from(async function*() {
-                for (let i = 0; i < data.length; i += CHUNK) {
-                    yield data.slice(i, i + CHUNK);
+                for (let i = 0; i < data.length; i += chunk_size) {
+                    yield data.slice(i, i + chunk_size);
                 }
             }());
             const chunk_fs = new ChunkFS({
@@ -134,8 +136,10 @@ async function file_target() {
                 `DataOriginMD5=${content_md5}`,
             );
             assert.strictEqual(content_md5, write_hash);
-            assert.strictEqual(chunk_fs.digest, content_md5);
-            assert.strictEqual(chunk_fs.digest, write_hash);
+            if (config.NSFS_CALCULATE_MD5) {
+                assert.strictEqual(chunk_fs.digest, content_md5);
+                assert.strictEqual(chunk_fs.digest, write_hash);
+            }
             // Leave parts on error
             fs.rmSync(F_TARGET);
         } finally {

--- a/src/util/chunk_fs.js
+++ b/src/util/chunk_fs.js
@@ -26,6 +26,8 @@ class ChunkFS extends stream.Transform {
         this.rpc_client = rpc_client;
         this.namespace_resource_id = namespace_resource_id;
         this._total_num_buffers = 0;
+        const platform_iov_max = nb_native().fs.PLATFORM_IOV_MAX;
+        this.iov_max = platform_iov_max ? Math.min(platform_iov_max, config.NSFS_DEFAULT_IOV_MAX) : config.NSFS_DEFAULT_IOV_MAX;
     }
 
     async _transform(chunk, encoding, callback) {
@@ -44,8 +46,9 @@ class ChunkFS extends stream.Transform {
                 const buf = (available_size < chunk.length) ? chunk.slice(0, available_size) : chunk;
                 this.q_buffers.push(buf);
                 this.q_size += buf.length;
-                // Should flush when equals, but added greater than just in case
-                if (this.q_size >= config.NSFS_BUF_SIZE) await this._flush_buffers();
+                // Should flush when num of chunks equals to max iov which is the limit according to https://linux.die.net/man/2/writev
+                // or when q_size equals to config.NSFS_BUF_SIZE, but added greater than just in case
+                if (this.q_buffers.length === this.iov_max || this.q_size >= config.NSFS_BUF_SIZE) await this._flush_buffers();
                 chunk = (available_size < chunk.length) ? chunk.slice(available_size) : null;
             }
             return callback();

